File: noNonsense/02_sliding_window/active_recall.java
// --- SLIDING WINDOW ACTIVE RECALL ---

public class SlidingWindow {
    // Type A: Fixed Size
    public int fixedWindow(int[] arr, int k) {
        int currSum = 0;
        for (int i = 0; i < ___; i++) currSum += arr[i];
        int maxSum = currSum;

        for (int i = 0; i < arr.length - ___; i++) {
            currSum = currSum - arr[___] + arr[i + ___];
            maxSum = Math.max(maxSum, ___);
        }
        return maxSum;
    }

    // Type B: Variable Size
    public int variableWindow(int[] arr, int target) {
        int left = 0, currSum = 0, res = Integer.MAX_VALUE;
        for (int right = 0; right < arr.length; right++) {
            currSum += arr[___];
            while (___ >= target) {
                res = Math.min(res, ___ - ___ + 1);
                currSum -= arr[___];
                ___++;
            }
        }
        return res == Integer.MAX_VALUE ? 0 : res;
    }
}

File: noNonsense/02_sliding_window/active_recall.py
# --- SLIDING WINDOW ACTIVE RECALL ---

# Type A: Fixed Size
def fixed_window(arr, k):
    # 1. Initialize first window
    curr_sum = sum(arr[:___])
    max_sum = curr_sum
    
    # 2. Slide
    for i in range(len(arr) - ___):
        curr_sum = curr_sum - arr[___] + arr[i + ___]
        max_sum = max(max_sum, ___)
    
    return max_sum

# Type B: Variable Size (Shrinkable)
def variable_window(arr, target):
    left = 0
    curr_sum = 0
    res = float('inf')
    
    for right in range(len(arr)):
        # 1. Expand
        curr_sum += arr[___]
        
        # 2. Shrink while valid
        while ___ >= target:
            res = min(res, ___ - ___ + 1)
            curr_sum -= arr[___]
            ___ += 1
            
    return res if res != float('inf') else 0
